Add initial support for i.MX8 platform

dbaluta opened the pull request:
This is patch series adds the basic infrastructure to support i.MX8 platforms. First two patches are just some cleanup in order to correctly compile and boot the FW.

Next patches introduce xtensa configuration supported by Hifi4 DSP found on i.MX8 platforms, rimage support, basic platform support and skeleton drivers for DAI and DMA.

I want to get the basic patches in because SOF is rapidly changing and I always spend times on fixing the rebase. 

After this gets in I will send the implementation of the drivers and topology files which will make SOF on i.MX8 usable.

tlauda commented on src/arch/xtensa/exc-dump.S:
```diff
@@ -29,6 +29,7 @@
 
 #include <arch/offset-defs.h>
 #include <xtensa/corebits.h>
+#include <config.h>
```
Not needed.

dbaluta replied:
> Not needed.

@tlauda Indeed. Initially, I've added this because I was using CONFIG_IMX8 but then I noticed the XCHAL_INTLEVEL6_MASK. Will fix in v2.

tlauda commented on src/platform/imx8/include/arch/xtensa/config/core-isa.h:
```diff
@@ -0,0 +1,626 @@
+   CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+   TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+   SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.  */
+
```
Are you using xcc for building?
We tend to add another guard here to detect wrong inclusions:
```
#if !defined __XCC__
...
#else

#error "xcc should not use this header"

#endif /* __XCC__ */
```

dbaluta replied:
@tlauda for the moment I'm using gcc but I will need to also support xcc. I get your point will fix it in v2.

tlauda commented on src/platform/imx8/dma.c:
```diff
@@ -0,0 +1,51 @@
+extern struct dma_ops dummy_dma_ops;
+extern struct dma_ops edma_ops;
+
+struct dma dma[2] = {
```
Some descriptive define would be better than magic number.

tlauda commented on src/platform/imx8/include/platform/clk-map.h:
```diff
@@ -0,0 +1,20 @@
+	{666000000, 666000, 0x0}, /* default */
+};
+
+static const struct freq_table ssp_freq[] = {
```
If you don't have SSP, why do you keep this?
Probably it's not abstracted enough to build without it, so maybe we should mark it as TODO.

dbaluta replied:
@tlauda yes, exactly. Will mark it as TODO.

lgirdwood replied:
Fwiw, I think this is what @jgunn0262 was working on in a recent PR, i.e. making this generic since SSP is Intel only.

plbossart commented on src/arch/xtensa/up/xtos/reset-vector.S:
```diff
@@ -86,7 +86,8 @@ _ResetVector:
 	|| defined (CONFIG_HASWELL) || defined(CONFIG_BROADWELL) \
-	|| defined(CONFIG_SKYLAKE) || defined(CONFIG_KABYLAKE)
+	|| defined(CONFIG_SKYLAKE) || defined(CONFIG_KABYLAKE) \
+	|| defined(CONFIG_IMX8)
```
this is starting to get ugly, maybe we should have an internal define that's set by the different platforms.

lgirdwood replied:
yep, something in Kconfig automatically selected by platform.

plbossart commented on src/platform/imx8/memory.c:
```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright 2019 NXP
+ *
```
new file, need a license !

I now see the link with the SPDX question, now we need to agree on that *before* merging this code

plbossart commented on src/platform/imx8/dai.c:
```diff
@@ -0,0 +1,34 @@
+/*
+ * Copyright 2019 NXP
+ *
```
license needed

plbossart commented on src/platform/imx8/dma.c:
```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright 2019 NXP
+ *
```
license needed

plbossart commented on src/platform/imx8/include/platform/dai.h:
```diff
@@ -0,0 +1,11 @@
+/*
+ * Copyright 2019 NXP
+ *
```
license needed

plbossart commented on src/platform/imx8/include/platform/dma.h:
```diff
@@ -0,0 +1,20 @@
+/*
+ * Copyright 2019 NXP
+ *
```
license needed

plbossart commented on src/platform/imx8/include/platform/idc.h:
```diff
@@ -0,0 +1,17 @@
+/*
+ * Copyright 2019 NXP
+ *
```
license needed

plbossart commented on src/platform/imx8/include/platform/mailbox.h:
```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright 2019 NXP
+ *
```
license needed

plbossart commented on src/platform/imx8/include/platform/memory.h:
```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright 2019 NXP
+ *
```
license needed

plbossart commented on src/platform/imx8/include/platform/mu.h:
```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2019 NXP
+ * Author: Daniel Baluta <daniel.baluta@nxp.com>
```
license needed

plbossart commented on src/platform/imx8/include/platform/pm_runtime.h:
```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright 2019 NXP
+ *
```
license needed

plbossart commented on src/platform/imx8/include/platform/pmc.h:
```diff
@@ -0,0 +1,17 @@
+/*
+ * Copyright 2019 NXP
+ *
```
license needed

plbossart commented on src/platform/imx8/include/platform/timer.h:
```diff
@@ -0,0 +1,22 @@
+/*
+ * Copyright 2019 NXP
+ *
```
license needed

plbossart commented on src/platform/imx8/include/platform/timer.h:
```diff
@@ -0,0 +1,22 @@
+#define TIMER0		IRQ_NUM_TIMER0
+#define TIMER1		IRQ_NUM_TIMER1
+/* FIXME: we only support 2 timers */
+#define TIMER2		-1
```
does SOF require 3 timers? What's the assumption here?

dbaluta replied:
> does SOF require 3 timers? What's the assumption here?

At some point yes. I see now that we can use less than three timers after commit 36d834855e0 (" xtensa: timer: Add support for less than three xtensa timers").

Will fix in v2.

plbossart commented on src/platform/imx8/platform.c:
```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright 2019 NXP
+ *
```
license needed

jajanusz commented on src/platform/imx8/include/platform/interrupt.h:
```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright 2019 NXP
```
License needed ;)

jajanusz commented on src/platform/imx8/include/platform/platform.h:
```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright 2019 NXP
```
License needed.

jajanusz commented:
@lgirdwood @dbaluta 
Are we going to add this platform to our CI? It would be good to check if it builds at least.

dbaluta commented:
> @lgirdwood @dbaluta
> Are we going to add this platform to our CI? It would be good to check if it builds at least.



> @lgirdwood @dbaluta
> Are we going to add this platform to our CI? It would be good to check if it builds at least.

Yes, as far as I understand I need to work with @xiulipan  for this to happen. The support for imx is already in the toolchain so we just need to add the commands for compilation.

lgirdwood commented on src/drivers/imx/pmc-ipc.c:
```diff
@@ -0,0 +1,33 @@
+/*
```
This file is probably not needed unless you need to perform IPC with an external PMIC.

dbaluta replied:
> This file is probably not needed unless you need to perform IPC with an external PMIC.

Indeed. Will fix in v2.

dbaluta commented:
@plbossart @lgirdwood @jajanusz @tlauda forced pushed new version. I think I addressed all of your comments except refactoring of  ("reset: xtos: Add support for i.mx8 reset flow").

A tentative patch for this refactoring is:
```

 /* Apollolake+ have reset vector in ROM */
-#if defined(CONFIG_BAYTRAIL) || defined(CONFIG_CHERRYTRAIL) \
-       || defined (CONFIG_HASWELL) || defined(CONFIG_BROADWELL) \
-       || defined(CONFIG_SKYLAKE) || defined(CONFIG_KABYLAKE) \
-       || defined(CONFIG_IMX8)
+#if defined(CONFIG_HAVE_RESET_VECTOR_ROM)
        j       _ResetHandler
 #else
 
```

The problem here is that I'm not sure if the naming is well chosen.

lgirdwood commented:
@dbaluta CI showing that your logic is OK :)

dbaluta commented:
> @dbaluta CI showing that your logic is OK :) EDIT - ok, I see this is not in the PR yet, best to add it as CI will tell you if logic is good.

By logic I mean: CONFIG_HAVE_RESET_VECTOR_ROM is the correct name for the macro. That is I wasn't sure that appollolake+ means: baytrail, cherrytrail, haswell, broadwell, skylake and kabylake. But I think that's fine.

My proposal is to send this fix as a separate PR after the current PR is merged.

lgirdwood commented:
@jajanusz @plbossart @tlauda is the latest update good for you guys, if so please approve.

jajanusz commented on src/drivers/CMakeLists.txt:
```diff
@@ -1,2 +1,6 @@
+if(CONFIG_IMX8)
```
Probably we should ifdef also intel & dw stuff later
